src/redis_cache.cc

```diff
@@ -0,0 +1,443 @@
+    options.host = host.substr(0, colon_pos);
+    options.port = std::stoi(host.substr(colon_pos + 1));
+  }
+  options.user = user_name;
+  options.password = password;
+  options.keep_alive = true;
+
+  sw::redis::ConnectionPoolOptions pool_options;
+  pool_options.size = 1;
+
+  // Connect to cluster.
+  std::cout << "Connecting via " << options.host << ':' << options.port << "..." << std::endl;
+  std::unique_ptr<sw::redis::Redis> redis = std::make_unique<sw::redis::Redis>(options, pool_options);
+  return redis;
+}
+
+void cache_set(
+  std::unique_ptr<sw::redis::Redis> &redis,
+  RedisCacheEntry<std::string> &cache_entry) {
+
+    //use hash_tag here
+    const std::string& prefix = "{" + cache_entry.key + "}";
+
+    // set number of entries in the top level
+    const std::string& entries_v = std::to_string(cache_entry.num_entries);
+
+    //set num entries prefix
+    bool is_set = redis->set(prefix, entries_v);
+    std::cout << std::to_string(is_set) << std::endl;
+
+
+    int i = 1;
+    for (auto &output: cache_entry.outputs) {
+      const std::string& key_v = prefix + "." + std::to_string(i);
+
+      // set response in a redis hash field
+      redis->hmset(key_v, output.fields.begin(), output.fields.end());
+      i += 1;
+    }
+}
+
+
+RedisCacheEntry<std::string> cache_get(
+    std::unique_ptr<sw::redis::Redis> &redis,
+    const std::string& hash_key) {
+
+    RedisCacheEntry<std::string> entry;
+
+    //use hash_tag here
+    const std::string& prefix = "{" + hash_key + "}";
+
+    std::string num = redis->get(prefix).value();
+    int num_entries = std::atoi(num.c_str());
+
+    std::vector<HashEntry<std::string>> entries;
+
+    /// TODO pipeline these
+    for (int i=1; i <=num_entries; i++) {
+      HashEntry<std::string> h;
+      const std::string& key = prefix + "." + std::to_string(i);
+      redis->hgetall(key, std::inserter(h.fields, h.fields.begin()));
+      entries.emplace_back(h);
+    }
+
+    entry.key = prefix;
+    entry.num_entries = num_entries;
+    entry.outputs = entries;
+
+    return entry;
+}
+
+
+
+namespace triton { namespace core {
+
+
+
+Status
+RedisResponseCache::Create(
+    std::string address,
+    std::string username,
+    std::string password,
+    std::unique_ptr<RequestResponseCache>* cache)
+{
+  try {
+    cache->reset(new RedisResponseCache(address, username, password));
+  }
+  catch (const std::exception& ex) {
+    return Status(
+        Status::Code::INTERNAL,
+        "Failed to initialize Response Cache: " + std::string(ex.what()));
+  }
+
+  return Status::Success;
+}
+
+RedisResponseCache::RedisResponseCache(std::string address, std::string username, std::string password)
+{
+
+  try {
+    this->_client = init_client(address, username, password);
+  }
+  catch (const std::exception& ex) {
+    throw std::runtime_error(
+        "Failed to initialize Redis Response Cache: " + std::string(ex.what()));
+  }
+
+  LOG_INFO << "Redis Response Cache is located at:" << address;
```

Comment:
What are the main types of communication expected with Redis? For `localhost` I think it's OK, but for a `remote` redis connection, what types of protocols are used? Would security-conscious users be concerned with the transmission of data? 

In general, are there further knobs to turn and configure for communicating with Redis, or is `address/username/password` all a user would ever need to connect? Would more configurable options around the network communication or performance need to be exposed here, or would that all be done on the `redis` server setup side? 

Any links to relevant documentation would be great as well.

Reply:
Three main types of communication
- normal TCP (with or w/o ACL security)
- TLS (private public key cert)
- UNIX (recommended if triton and redis are co-located on the same server)


My plan is actually to take out the hooks for the current connection string and add in the ability to just pass a JSON file. 

so for ex. for TLS, you would pass in file paths to cert and key files. For ACL/TCP you would pass in username/password. for unix, a unix connection string. All of which will be specified in a JSON file.

this way, if anything ever needs to be added or removed, or other variants down the road get added, it'll be easy to add in.

here are the [docs for redis-plus-plus](https://github.com/sewenew/redis-plus-plus#connection) connections.